api/v1beta1/grpc_disruption.go

```diff
@@ -0,0 +1,146 @@
+	// +ddmark:validation:Enum=OK;CANCELED;UNKNOWN;INVALID_ARGUMENT;DEADLINE_EXCEEDED;NOT_FOUND;ALREADY_EXISTS;PERMISSION_DENIED;RESOURCE_EXHAUSTED;FAILED_PRECONDITION;ABORTED;OUT_OF_RANGE;UNIMPLEMENTED;INTERNAL;UNAVAILABLE;DATA_LOSS;UNAUTHENTICATED
+	ErrorToReturn string `json:"error,omitempty"`
+	// +kubebuilder:validation:Enum={}
+	// +ddmark:validation:Enum={}
+	OverrideToReturn string `json:"override,omitempty"`
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=100
+	// +ddmark:validation:Minimum=1
+	// +ddmark:validation:Maximum=100
+	QueryPercent int `json:"query_pct,omitempty"`
+}
+
+const (
+	ERROR    = "error"
+	OVERRIDE = "override"
+)
+
+// Validate validates that there are no missing hostnames or records for the given grpc disruption spec
+func (s GRPCDisruptionSpec) Validate() error {
+	if s.Port == 0 {
+		return errors.New("the gRPC disruption does not specify the port which the target service exposes, but port must be specified")
+	}
+
+	if len(s.Endpoints) == 0 {
+		return errors.New("the gRPC disruption was selected with no endpoints specified, but endpoints must be specified")
+	}
+
+	queryPctByEndpoint := make(map[string]int)
+
+	for _, alteration := range s.Endpoints {
+		if alteration.TargetEndpoint == "" {
+			return errors.New("some list items in gRPC disruption are missing endpoints; specify a TargetEndpoint for each item in the list")
+		}
+
+		// check that endpoint is not already configured such that the sum of mangled queryPercents total to more than 100%
+		if totalQueryPercent, ok := queryPctByEndpoint[alteration.TargetEndpoint]; ok {
+			if alteration.QueryPercent > 0 {
+				if totalQueryPercent+alteration.QueryPercent > 100 {
+					return errors.New("total queryPercent of all alterations applied to endpoint %s is over 100%; modify them to so their total is 100% or less")
+				}
+
+				queryPctByEndpoint[alteration.TargetEndpoint] += alteration.QueryPercent
+			}
+		} else {
+			if alteration.QueryPercent > 100 {
+				return errors.New("total queryPercent of all alterations applied to endpoint %s is over 100%; modify them to so their total is 100% or less")
+			}
+
+			queryPctByEndpoint[alteration.TargetEndpoint] = alteration.QueryPercent
+		}
+
+		// check that exactly one of ErrorToReturn or OverrideToReturn is configured
+		if alteration.ErrorToReturn != "" && alteration.OverrideToReturn != "" {
+			return fmt.Errorf("the gRPC disruption has ErrorToReturn and OverrideToReturn specified for endpoint %s, but it can only have one", alteration.TargetEndpoint)
+		} else if alteration.ErrorToReturn == "" && alteration.OverrideToReturn == "" {
+			return fmt.Errorf("the gRPC disruption must have either ErrorToReturn or OverrideToReturn specified for endpoint %s", alteration.TargetEndpoint)
+		}
+
+		if alteration.ErrorToReturn != "" && alteration.ErrorToReturn != "OK" &&
```

Comment:
I think we should be using the error map `grpc/server.go` here, rather than re-listing every error type explicitly, wdyt?

Reply:
there is an import cycle 😭 
![image](https://user-images.githubusercontent.com/4872808/133844792-9050d0a1-3a3e-48c6-b5e4-c0fbf86120ef.png)

Reply:
oh I guess I could move it?

Reply:
I moved the map so there's no cycle